File: generative-ui/backend/app/services/geocoding.py
"""
Dynamic geocoding service using Nominatim (OpenStreetMap)
Free, no API key required, rate limit: 1 request/second
"""
import httpx
import asyncio
import re
from typing import Optional, Dict
from datetime import datetime


class GeocodingService:
    """
    Free geocoding using Nominatim (OpenStreetMap)
    No API key needed, rate limit: 1 request/second
    """
    
    BASE_URL = "https://nominatim.openstreetmap.org"
    
    def __init__(self):
        self.last_request_time = 0.0
        self.cache: Dict[str, Optional[Dict[str, float]]] = {}
        
    async def geocode(self, location: str) -> Optional[Dict[str, float]]:
        """
        Convert location string to coordinates
        
        Args:
            location: "Tokyo, Japan" or "123 Main St, San Francisco" or "Marina Bay Sands"
            
        Returns:
            {"lat": 35.6762, "lng": 139.6503, "display_name": "Tokyo, Japan"} or None
        """
        # Check cache first
        if location in self.cache:
            print(f"[Geocoding] Cache hit for: {location}")
            return self.cache[location]
        
        try:
            # Respect rate limit (1 req/sec for Nominatim)
            await self._rate_limit()
            
            print(f"[Geocoding] Requesting coordinates for: {location}")
            
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.BASE_URL}/search",
                    params={
                        "q": location,
                        "format": "json",
                        "limit": 1,
                        "addressdetails": 1
                    },
                    headers={
                        "User-Agent": "GenerativeUIBrowser/1.0 (Educational Project)"
                    },
                    timeout=10.0
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data and len(data) > 0:
                        result = data[0]
                        coords = {
                            "lat": float(result["lat"]),
                            "lng": float(result["lon"]),
                            "display_name": result.get("display_name", location)
                        }
                        
                        print(f"[Geocoding] ✓ Found: {coords['lat']}, {coords['lng']}")
                        
                        # Cache result
                        self.cache[location] = coords
                        return coords
                    else:
                        print(f"[Geocoding] ✗ No results for: {location}")
                else:
                    print(f"[Geocoding] ✗ HTTP {response.status_code} for: {location}")
                        
        except Exception as e:
            print(f"[Geocoding] ✗ Error for '{location}': {e}")
        
        # Cache failure to avoid retry
        self.cache[location] = None
        return None
    
    async def reverse_geocode(self, lat: float, lng: float) -> Optional[str]:
        """
        Convert coordinates to location string
        
        Args:
            lat: 35.6762
            lng: 139.6503
            
        Returns:
            "Tokyo, Japan" or None
        """
        cache_key = f"{lat},{lng}"
        if cache_key in self.cache:
            cached = self.cache[cache_key]
            return cached.get("display_name") if cached else None
        
        try:
            await self._rate_limit()
            
            print(f"[Reverse Geocoding] Requesting address for: {lat}, {lng}")
            
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.BASE_URL}/reverse",
                    params={
                        "lat": lat,
                        "lon": lng,
                        "format": "json"
                    },
                    headers={
                        "User-Agent": "GenerativeUIBrowser/1.0 (Educational Project)"
                    },
                    timeout=10.0
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if "display_name" in data:
                        display_name = data["display_name"]
                        print(f"[Reverse Geocoding] ✓ Found: {display_name}")
                        
                        # Cache result
                        self.cache[cache_key] = {
                            "lat": lat,
                            "lng": lng,
                            "display_name": display_name
                        }
                        return display_name
                        
        except Exception as e:
            print(f"[Reverse Geocoding] ✗ Error for ({lat}, {lng}): {e}")
        
        self.cache[cache_key] = None
        return None
    
    async def geocode_batch(self, locations: list[str]) -> Dict[str, Optional[Dict[str, float]]]:
        """
        Geocode multiple locations with rate limiting
        
        Args:
            locations: ["Tokyo", "Paris", "New York"]
            
        Returns:
            {"Tokyo": {"lat": 35.6762, "lng": 139.6503}, ...}
        """
        results = {}
        
        for location in locations:
            coords = await self.geocode(location)
            results[location] = coords
            
        return results
    
    async def _rate_limit(self):
        """
        Enforce 1 request per second rate limit for Nominatim
        """
        current_time = asyncio.get_event_loop().time()
        time_since_last = current_time - self.last_request_time
        
        if time_since_last < 1.0:
            wait_time = 1.0 - time_since_last
            print(f"[Geocoding] Rate limiting: waiting {wait_time:.2f}s")
            await asyncio.sleep(wait_time)
            
        self.last_request_time = asyncio.get_event_loop().time()
    
    def clear_cache(self):
        """Clear the geocoding cache"""
        self.cache.clear()
        print("[Geocoding] Cache cleared")


# Singleton instance
geocoding_service = GeocodingService()

File: generative-ui/frontend/src/utils/geocoding.ts
/**
 * Frontend geocoding using Nominatim (OpenStreetMap)
 * Only used as fallback if backend didn't provide coordinates
 * 
 * Usage:
 * const coords = await geocode("Tokyo, Japan");
 * // Returns: { lat: 35.6762, lng: 139.6503 } or null
 */

const NOMINATIM_URL = 'https://nominatim.openstreetmap.org';

// Cache to avoid repeated requests
const geocodeCache = new Map<string, { lat: number; lng: number } | null>();

// Rate limiting: last request timestamp
let lastRequestTime = 0;

/**
 * Geocode a location string to coordinates
 * Uses Nominatim API with rate limiting (1 req/sec)
 */
export async function geocode(location: string): Promise<{ lat: number; lng: number } | null> {
  console.log(`[Frontend Geocoding] Request for: "${location}"`);
  
  // Check cache
  if (geocodeCache.has(location)) {
    const cached = geocodeCache.get(location);
    console.log(`[Frontend Geocoding] Cache hit:`, cached);
    return cached || null;
  }
  
  try {
    // Rate limiting: wait at least 1 second between requests
    const now = Date.now();
    const timeSinceLast = now - lastRequestTime;
    
    if (timeSinceLast < 1000) {
      const waitTime = 1000 - timeSinceLast;
      console.log(`[Frontend Geocoding] Rate limiting: waiting ${waitTime}ms`);
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }
    
    lastRequestTime = Date.now();
    
    console.log(`[Frontend Geocoding] Calling Nominatim API...`);
    
    const response = await fetch(
      `${NOMINATIM_URL}/search?` + new URLSearchParams({
        q: location,
        format: 'json',
        limit: '1'
      }),
      {
        headers: {
          'User-Agent': 'GenerativeUIBrowser/1.0 (Educational Project)'
        }
      }
    );
    
    if (response.ok) {
      const data = await response.json();
      
      if (data && Array.isArray(data) && data.length > 0) {
        const result = data[0];
        const coords = {
          lat: parseFloat(result.lat),
          lng: parseFloat(result.lon)
        };
        
        console.log(`[Frontend Geocoding] ✓ Found:`, coords);
        
        // Cache result
        geocodeCache.set(location, coords);
        return coords;
      } else {
        console.warn(`[Frontend Geocoding] ✗ No results for: "${location}"`);
      }
    } else {
      console.error(`[Frontend Geocoding] ✗ HTTP ${response.status}`);
    }
  } catch (error) {
    console.error(`[Frontend Geocoding] ✗ Error:`, error);
  }
  
  // Cache failure to avoid retrying
  geocodeCache.set(location, null);
  return null;
}

/**
 * Reverse geocode coordinates to location string
 */
export async function reverseGeocode(lat: number, lng: number): Promise<string | null> {
  const cacheKey = `${lat},${lng}`;
  console.log(`[Frontend Reverse Geocoding] Request for: ${cacheKey}`);
  
  // Check cache
  if (geocodeCache.has(cacheKey)) {
    const cached = geocodeCache.get(cacheKey);
    return cached ? JSON.stringify(cached) : null;
  }
  
  try {
    // Rate limiting
    const now = Date.now();
    const timeSinceLast = now - lastRequestTime;
    
    if (timeSinceLast < 1000) {
      await new Promise(resolve => setTimeout(resolve, 1000 - timeSinceLast));
    }
    
    lastRequestTime = Date.now();
    
    const response = await fetch(
      `${NOMINATIM_URL}/reverse?` + new URLSearchParams({
        lat: lat.toString(),
        lon: lng.toString(),
        format: 'json'
      }),
      {
        headers: {
          'User-Agent': 'GenerativeUIBrowser/1.0 (Educational Project)'
        }
      }
    );
    
    if (response.ok) {
      const data = await response.json();
      
      if (data && data.display_name) {
        console.log(`[Frontend Reverse Geocoding] ✓ Found:`, data.display_name);
        return data.display_name;
      }
    }
  } catch (error) {
    console.error(`[Frontend Reverse Geocoding] ✗ Error:`, error);
  }
  
  return null;
}

/**
 * Batch geocode multiple locations
 * Uses rate limiting to avoid overwhelming API
 */
export async function geocodeBatch(locations: string[]): Promise<Map<string, { lat: number; lng: number } | null>> {
  console.log(`[Frontend Geocoding] Batch request for ${locations.length} locations`);
  
  const results = new Map<string, { lat: number; lng: number } | null>();
  
  for (const location of locations) {
    const coords = await geocode(location);
    results.set(location, coords);
  }
  
  console.log(`[Frontend Geocoding] Batch complete: ${results.size} results`);
  return results;
}

/**
 * Clear the geocoding cache
 */
export function clearGeocodeCache() {
  geocodeCache.clear();
  console.log('[Frontend Geocoding] Cache cleared');
}
